docs/when_internals.md

```diff
@@ -0,0 +1,342 @@
+    primitive instance.
+* `backend/mlir/when.py`
+  * `WhenCompiler`
+    * Overview
+      * A class encapsulating the MLIR compilation logic for instances of
+        the `WhenBuilder` primitive.  It is invoked by the `ModuleVisitor`
+        class in `backend/mlir/harware_module.py`.
+    * Attributes
+      * `_module_visitor`: handle to `ModuleVisitor` instance from
+        `backend/mlir/harware_module.py`, used to reference the context for
+        creating values, compile options (e.g. `flatten_all_tuples`), and value
+        creation helpers (e.g. `make_array_ref`).
+      * `_module`: reference to the when primitive MLIR module being compiled
+      * `_operands`: reference to `_module.operands`
+      * `_flatten_all_tuples`: reference to `flatten_all_tuples` compile option
+      * `_outer_block`: get the current active MLIR block, this is used to
+        create value references in the enclosing scope, which aims to reduce
+        code duplication (since references are cached per block).
+      * `_builder`: reference to the magma builder object corresponding to the
+        when instance being compiled
+      * `_input_to_index`, `_output_to_index`: copies of the builder index maps
+        where tuple values are flattened to match the MLIR `flatten_all_tuples`
+        logic
+      * `_output_wires`: a list of mlir values corresponding to each module
+        output
+    * Methods
+      * `_get_input_index`: retrieve an input magma value's index from the
+          `_input_to_index` map
+      * `_get_operand`: retrieve the MLIR operand value for a magma value
+        (using `_get_input_index` to lookup the position in `_operands`)
+      * `_flatten_index_map`: flattens tuples values in  the builder primitive
+        index maps to match the MLIR flatten tuples logic
+      * `_make_output_wires`: construct the `RegOp` instances for each of the
+        module results
+      * `_flatten_value`: given a magma value, flatten it into a list of
+        children (for flatten all tuples)
+      * `_get_parent`: traverse the ancestor chain for an array to see if it's
+        contained in the provided index map.  If so, use the index to lookup
+        the corresponding MLIR value in the provided collection and return
+        a reference to the corresponding array ancestor.  Used to pack
+        assignments for an array value.
+      * `_check_array_child_wire`: see if a value is a child of an array that is
+        in the index map.  If so, return the parent wire and an index
+        corresponding to the current child.
+      * `_make_operand`: If the operand is an element of wire, emit the
+        required array reference ops (extract, get, or slice) to retrieve the desired index.
+      * `_build_wire_map`: Collect a map of output wires to their drivers. If
+        it's an array that's been elaborated, we collect the drivers in a
+        dictionary using their index as a key to sort.
```

Comment:
I think there's some room for improvement for making the wire map logic more structured, but this is the core code responsible for mapping magma conditional wires to MLIR assignments